helix-core/src/main/java/org/apache/helix/controller/stages/IntermediateStateCalcStage.java

```diff
@@ -631,104 +631,53 @@ private void chargePendingTransition(Resource resource, CurrentStateOutput curre
       ResourceControllerDataProvider cache) {
-
-    Map<String, String> currentStateMap =
-        currentStateOutput.getCurrentStateMap(resourceName, partition);
-    Map<String, String> bestPossibleMap = bestPossiblePartitionStateMap.getPartitionMap(partition);
-    Set<String> allInstances = new HashSet<>(currentStateMap.keySet());
-    allInstances.addAll(bestPossibleMap.keySet());
-    Map<String, String> intermediateMap = new HashMap<>();
-
     boolean hasReachedThrottlingLimit = false;
     if (throttleController.shouldThrottleForResource(rebalanceType, resourceName)) {
       hasReachedThrottlingLimit = true;
       if (logger.isDebugEnabled()) {
-        LogUtil.logDebug(logger, _eventId,
-            String.format("Throttled on partition: %s in resource: %s",
-                partition.getPartitionName(), resourceName));
+        LogUtil.logDebug(logger, _eventId, String.format(
+            "Throttled because of cluster/resource quota is full for message {%s} on partition {%s} in resource {%s}",
+            messageToThrottle.getId(), partition.getPartitionName(), resourceName));
       }
     } else {
-      // throttle if any of the instances are not able to accept state transitions
-      for (String instance : allInstances) {
-        String currentState = currentStateMap.get(instance);
-        String bestPossibleState = bestPossibleMap.get(instance);
-        if (bestPossibleState != null && !bestPossibleState.equals(currentState)
-            && !cache.getDisabledInstancesForPartition(resourceName, partition.getPartitionName())
-                .contains(instance)) {
-          if (throttleController.shouldThrottleForInstance(rebalanceType, instance)) {
-            hasReachedThrottlingLimit = true;
-            if (logger.isDebugEnabled()) {
-              LogUtil.logDebug(logger, _eventId,
-                  String.format(
-                      "Throttled because of instance: %s for partition: %s in resource: %s",
-                      instance, partition.getPartitionName(), resourceName));
-            }
-            break;
+      // Since message already generated, we can assume the current state is not null and target state is not null
+      if (!cache.getDisabledInstancesForPartition(resourceName, partition.getPartitionName())
+          .contains(messageToThrottle.getTgtName())) {
+        if (throttleController.shouldThrottleForInstance(rebalanceType, messageToThrottle.getTgtName())) {
+          hasReachedThrottlingLimit = true;
+          if (logger.isDebugEnabled()) {
+            LogUtil.logDebug(logger, _eventId, String.format(
+                "Throttled because of instance level quota is full on instance {%s} for message {%s} of partition {%s} in resource {%s}",
+                messageToThrottle.getId(), messageToThrottle.getTgtName(), partition.getPartitionName(), resourceName));
           }
         }
       }
     }
+    // If there is a room to hold existing messages, charge this message.
```

Comment:
Nit: did you mean: "If there is still room for this replica, proceed to charge at the cluster and resource level and set the intermediate partition-state mapping so that the state transition will be generated."

Question: please confirm with me if my understanding is correct: we do not have `chargeReplica()`; instead we incorporate the replica-level throttling by having each message represent a replica.

So the number of calls for `throttleStateTransitionsForPartition()` would be the total number of partitions in the cluster whereas the number of calls for `throttleStateTransitionsForReplica()` would be the total number of replicas in the cluster, correct?

On an average case, we will have 3x replicas as partitions (because applications usually use 3 as the replication factor). I am curious, are you able to do a quick local benchmark of the `IntermediateStateCalcStage` for before & after for a production-grade cluster? It might help us see if we are unwittingly creating a bottleneck here somewhere.

Reply:
For your question:

1. Yes, even this function is called more times than single partition, there is no cost operations, such as new object creation / deletion, data collection looping. The only operations are read/update the map in throttle controllers and intermediate state map.
2. Usually you will have 3x as the one partition for throttling. Because a lot of messages have dependencies, which are removed by MessageSelectionStage. From historical data in production log, you can see that the number of messages is far less than what we have now.

Finally, as I said, this is non avoidable operation. Because each of the throttling charge operation success or fail can lead different result of next message rebalance type. The only thing we can do is trying to make sure no cost operations for this simple function call and that is also the reason I am trying to let this function to be reviewed independently instead of combing it into a big logic code.

Reply:
@dasahcc 

I agree that if we want to do replica-based throttling, this is something that would be needed considering the way we throttle. 

Have we, though, considered making this optional/configurable? For example, many users of Helix do not require throttling capability at this level of granularity?

Reply:
That's a good suggestion! If user does not enable the throttling, we can skip this stage even. I will add this to following PR.